web-console/src/lib/components/streaming/management/AnalyticsPipelineTput.tsx

```diff
@@ -3,33 +3,49 @@
 // Also smooths the throughput over a few seconds.
 
 import { discreteDerivative } from '$lib/functions/common/math'
+import { tuple } from '$lib/functions/common/tuple'
 import { GlobalMetrics } from '$lib/types/pipeline'
 import { ApexOptions } from 'apexcharts'
 import { format } from 'd3-format'
 import ReactApexcharts from 'src/@core/components/react-apexcharts'
 
+import { Box } from '@mui/material'
 import Card from '@mui/material/Card'
-import CardContent from '@mui/material/CardContent'
 import { useTheme } from '@mui/material/styles'
 import Typography from '@mui/material/Typography'
 
-const AnalyticsPipelineTput = (props: { metrics: GlobalMetrics[] }) => {
+export const AnalyticsPipelineTput = (props: {
+  metrics: { global: (GlobalMetrics & { timeMs: number })[] }
+  keepMs: number
+}) => {
   const theme = useTheme()
 
-  const totalProcessed = discreteDerivative(props.metrics, m => m.total_processed_records).filter(x => x != 0)
-  const throughput = discreteDerivative(totalProcessed, (n1, n0) => n1 - n0)
-  const smoothTput = discreteDerivative(throughput, (n1, n0) => n1 * 0.6 + 0.4 * n0)
-
+  const totalProcessed = props.metrics.global.map(m => tuple(m.timeMs, m.total_processed_records))
+  const throughput = discreteDerivative(totalProcessed, (n1, n0) =>
+    tuple(n1[0], ((n1[1] - n0[1]) * 1000) / (n1[0] - n0[0]))
+  )
+  const smoothTput = ((n1, n0) => n1[1] * 0.5 + n0[1] * 0.5)(
+    throughput.at(-2) ?? throughput.at(-1) ?? tuple(0, 0),
+    throughput.at(-1) ?? tuple(0, 0)
+  )
   const series = [
     {
       data: throughput
     }
   ]
 
+  const valueMax = throughput.length ? Math.max(...throughput.map(v => v[1])) : 0
```

Comment:
all these MUI tools and you have to do this by hand? there's no plotting tool which knows how to scale automatically?

Reply:
I am using ApexCharts library for this; it does have auto-scaling, but it doesn't default to nice rounded min and max numbers, so I implemented this myself